test/e2e/deployment.go

```diff
@@ -270,3 +274,131 @@ func testRollingUpdateDeploymentEvents(f *Framework) {
+	defer func() {
+		Logf("deleting replication controller %s", rcName)
+		Expect(c.ReplicationControllers(ns).Delete(rcName)).NotTo(HaveOccurred())
+	}()
+	// Verify that the required pods have come up.
+	err = verifyPods(c, ns, "sample-pod", false, rcReplicas)
+	if err != nil {
+		Logf("error in waiting for pods to come up: %s", err)
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	// Create a deployment to delete nginx pods and instead bring up redis-slave pods.
+	deploymentName, deploymentImageName := "redis-deployment", "redis-slave"
+	deploymentReplicas := 4
+	deploymentImage := "gcr.io/google_samples/gb-redisslave:v1"
+	deploymentMinReadySeconds := 5
+	Logf("Creating deployment %s", deploymentName)
+	newDeployment := extensions.Deployment{
+		ObjectMeta: api.ObjectMeta{
+			Name: deploymentName,
+		},
+		Spec: extensions.DeploymentSpec{
+			Replicas:       deploymentReplicas,
+			Selector:       deploymentPodLabels,
+			UniqueLabelKey: "deployment.kubernetes.io/podTemplateHash",
+			Template: api.PodTemplateSpec{
+				ObjectMeta: api.ObjectMeta{
+					Labels: deploymentPodLabels,
+				},
+				Spec: api.PodSpec{
+					Containers: []api.Container{
+						{
+							Name:  deploymentImageName,
+							Image: deploymentImage,
+						},
+					},
+				},
+			},
+			Strategy: extensions.DeploymentStrategy{
+				RollingUpdate: &extensions.RollingUpdateDeployment{
+					MaxUnavailable:  intstr.FromInt(1),
+					MaxSurge:        intstr.FromInt(1),
+					MinReadySeconds: deploymentMinReadySeconds,
+				},
+			},
+		},
+	}
+	_, err = c.Deployments(ns).Create(&newDeployment)
+	Expect(err).NotTo(HaveOccurred())
+	defer func() {
+		deployment, err := c.Deployments(ns).Get(deploymentName)
+		Expect(err).NotTo(HaveOccurred())
+		Logf("deleting deployment %s", deploymentName)
+		Expect(c.Deployments(ns).Delete(deploymentName, nil)).NotTo(HaveOccurred())
+		// TODO: remove this once we can delete rcs with deployment
+		newRC, err := deploymentutil.GetNewRC(*deployment, c)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(c.ReplicationControllers(ns).Delete(newRC.Name)).NotTo(HaveOccurred())
+	}()
+	// Verify that the pods were scaled up and down as expected. We use events to verify that.
+	deployment, err := c.Deployments(ns).Get(deploymentName)
+	// Make sure the deployment starts to scale up and down RCs
+	waitForPartialEvents(c, ns, deployment, 2)
+	events, err := c.Events(ns).Search(deployment)
+	if err != nil {
+		Logf("error in listing events: %s", err)
+		Expect(err).NotTo(HaveOccurred())
+	}
+	newRC, err := deploymentutil.GetNewRC(*deployment, c)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(newRC).NotTo(Equal(nil))
+
+	// Before the deployment finishes, update the deployment to rollover the above 2 rcs and bring up redis pods.
+	// If the deployment already finished here, the test would fail. When this happens, increase its minReadySeconds or replicas to prevent it.
+	Expect(len(events.Items)).Should(BeNumerically("<", deploymentReplicas))
```

Comment:
I will suggest keeping the desired replicas very high like 10 or 20 rather than 4 to ensure that this never fails.
In newDeployment, you can change desired replicas to something small like 4.